bindings/rust/src/lib.rs

```diff
@@ -289,6 +289,76 @@ impl TypedExecutionResult {
 }
 
 impl Instance {
+    /// Ensure the range is valid according to the currently available memory size.
+    fn checked_memory_range(
+        &self,
+        offset: u32,
+        size: usize,
+    ) -> Result<core::ops::Range<usize>, ()> {
+        // This should be safe given usize::BITS >= u32::BITS, see https://doc.rust-lang.org/std/primitive.usize.html.
+        let offset = offset as usize;
+        let memory_size = self.memory_size();
+        // Empty slices are allowed, but ensure both starting and ending offsets are valid.
+        if offset >= memory_size || (offset + size) > memory_size {
+            return Err(());
+        }
+        // Assert this too?
+        // debug_assert!(memory_size != 0);
```

Comment:
can be uncommented

Reply:
Clarified that https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html allows `len=0`, just that `data` has to be valid.